snuba/query/processors/mapping_optimizer.py

```diff
@@ -159,20 +185,110 @@ def __replace_with_hash(self, condition: Expression) -> Expression:
+                          ^                            ^
+                          |                            |
+                      existence check               value check
+
+        the existence check in this clause is redundant and prevents the hashmap
+        optimization from being applied.
+
+        This function will remove all tag existence checks
+        from the condition IFF they are ANDed with a value check for the *same tag name*
+
+        Side effects:
+            This function works by flattening first level AND conditions to find clauses where
+            existence checks and value checks are ANDed together. When the AND conditions are recombined,
+            they are not guaranteed to be in the same structure (but are guaranteed to be functionally equivalent)
+
+            Example:
+                ┌───┐         ┌───┐
+                │AND│         │AND│
+                ├──┬┘         └┬──┤
+                │  │           │  │
+             ┌──┴┐ c           a ┌┴──┐
+             │AND│    becomes    │AND│
+             └┬─┬┘               ├──┬┘
+              │ │                │  │
+              a b                b  c
+        """
+        if not isinstance(condition, FunctionExpr):
+            return condition
+        elif condition.function_name == BooleanFunctions.OR:
+            sub_conditions = get_first_level_or_conditions(condition)
+            pruned_conditions = [
+                self._get_condition_without_redundant_checks(c, query)
+                for c in sub_conditions
+            ]
+            return combine_or_conditions(pruned_conditions)
+        elif condition.function_name == BooleanFunctions.AND:
+            sub_conditions = get_first_level_and_conditions(condition)
+            tag_eq_match_strings = set()
+            matched_tag_exists_conditions = {}
+            for condition_id, cond in enumerate(sub_conditions):
+                tag_exist_match = _tag_exists_pattern.match(cond)
+                if tag_exist_match:
+                    matched_tag_exists_conditions[condition_id] = tag_exist_match
+                eq_match = self.__optimizable_pattern.match(cond)
+                if eq_match:
+                    tag_eq_match_strings.add(eq_match.string("key"))
+            useful_conditions = []
+            for condition_id, cond in enumerate(sub_conditions):
+                tag_exist_match = matched_tag_exists_conditions.get(condition_id, None)
+                if tag_exist_match:
+                    requested_tag = tag_exist_match.string("key")
+                    if requested_tag in tag_eq_match_strings:
+                        # how often we should skip this optimization
+                        skip_rate = state.get_config(
+                            "tags_redundant_optimizer_skip_rate", 0
+                        )
+                        if isinstance(skip_rate, float) or isinstance(skip_rate, int):
```

Comment:
This is going to suffer from a mistake I made: you want to optimize the entire query or none of it. In this case, if someone has a check on two different tag keys like `(has(tags, A) AND tags[A] = 1) AND (has(tags, B) AND tags[B] = 1)` you might optimize the condition on A but not the one on B and end up with a query that is half optimized. This should also apply across the WHERE and HAVING clause so that either both are optimized or neither are.